compiler/crates/relay-lsp/src/rename/mod.rs

```diff
@@ -0,0 +1,416 @@
+        text_document: params.text_document_position.text_document,
+        position: params.text_document_position.position,
+    };
+    let (feature, position_span, source_location_key) =
+        state.extract_feature_from_text(&text_document_position_params, 1)?;
+
+    let program = &state.get_program(&state.extract_project_name_from_url(&uri)?)?;
+    let root_dir = &state.root_dir();
+
+    let changes = match feature {
+        crate::Feature::GraphQLDocument(document) => {
+            let node_path = document.resolve((), position_span);
+
+            match node_path {
+                ResolutionPath::Ident(IdentPath {
+                    inner: fragment_spread_name,
+                    parent: IdentParent::FragmentSpreadName(_),
+                }) => Ok(rename_fragment(
+                    fragment_spread_name.value,
+                    params.new_name,
+                    program,
+                    root_dir,
+                )),
+                ResolutionPath::Ident(IdentPath {
+                    inner: fragment_name,
+                    parent: IdentParent::FragmentDefinitionName(_),
+                }) => Ok(rename_fragment(
+                    fragment_name.value,
+                    params.new_name,
+                    program,
+                    root_dir,
+                )),
+                ResolutionPath::Ident(IdentPath {
+                    inner: operation_name,
+                    parent: IdentParent::OperationDefinitionName(_),
+                }) => {
+                    let location = common::Location::new(source_location_key, operation_name.span);
+
+                    let lsp_location =
+                        transform_relay_location_to_lsp_location(root_dir, location).unwrap();
+
+                    Ok(rename_operation(params.new_name, lsp_location))
+                }
+                _ => Err(LSPRuntimeError::ExpectedError),
+            }
+        }
+        crate::Feature::DocblockIr(docblock) => {
+            let resolution_info =
+                create_docblock_resolution_info(&docblock, position_span).unwrap();
+
+            match resolution_info {
+                DocblockResolutionInfo::FieldName(docblock_field) => {
+                    let parent_type = extract_parent_type(docblock);
+
+                    let mut changes = rename_relay_resolver_field(
+                        docblock_field.value,
+                        parent_type,
+                        &params.new_name,
+                        program,
+                        root_dir,
+                    );
+
+                    let location = common::Location::new(source_location_key, docblock_field.span);
+                    let lsp_location =
+                        transform_relay_location_to_lsp_location(root_dir, location)?;
+
+                    merge_text_edit(&mut changes, lsp_location, &params.new_name);
+
+                    // todo: rename JS function
```

Comment:
I think it would be really cool, if we could also rename the JS function of the Relay Resolver field, but I think the location of the function name isn't yet tracked by the LSP. I could probably hack together some lookahead after the DocBlock to find the function definition, but I don't think it would be very stable 🤪

Reply:
I think we'll have to leave that out for now. Eventually I would love Relay to understand the surrounding JavaScript (and types!) and to be able to infer a lot of what is currently duplicated between the docblock and the actual code (similar to what I'm exploring with [Grats](https://grats.capt.dev/)). However, that is a larger project.